### PR TITLE
Added CVE-2016-0040 Windows Privilege Escalation Exploit.

### PR DESCRIPTION
**Pull request by de7ec7ed**
This pull request adds a local privilege escalation exploit and associated metasploit module for CVE-2016-0040 to the project. It has been tested on Windows 7 SP0 and SP1 x86-64. It should be noted that this exploit does NOT require KASLR, SMEP or SMAP to be disable as it focuses on leveraging ring-0 as a confused deputy. Furthermore, ROP is NOT used and no shellcode is injected into or ran at ring-0. Nothing in this exploit is known to trigger PatchGuard BSoDs.

## Verification

List the steps needed to make sure this thing works

- [x] Start `msfconsole`
- [x] `use exploit/windows/local/ms16_014_wmi_recv_notif.rb`

### CODE REVIEW

**Comment by busterb** [comment]
Hi, thanks @de7ec7ed! Can you please add the source for the injected DLL?

**Comment by de7ec7ed** [comment]
Of course, I'll try to get the source code added this weekend. Until then if you want to review what I will be moving over here is the source code (https://github.com/de7ec7ed/CVE-2016-0040). I have the referenced repository running with a local CI and a vulnerable VM but am curious what I should do for this project once moved.

**Comment by de7ec7ed** [comment]
@busterb does the source code have to build with mingw or can i just add the visual studio project. It seems the 64bit reflective dll is having some issues with the mingw toolchain.

**Comment by bcook-r7** [comment]
the visual studio project is fine!

**Comment by de7ec7ed** [comment]
done.

**Comment by wwebb-r7** [comment]
A couple of things to start with:

- The DLL filename should reference the architecture it's built for, as in `exploit_name.(x86 | x64).dll`.  You have this correct in the module, but not in the actual file itself.  Your VS project output configuration should reflect this as well
- Source code goes in `external/source/exploits/exploit_name_or_cve_number`

**Comment by de7ec7ed** [comment]
done. I also recently tested this on windows server 2008 R2 SP1 can you tell me the best way to update the ruby file?

**Comment by busterb** [comment]
I think you did it!

**Comment by wwebb-r7** [comment]
Needs module documentation.  I should have mentioned it earlier but I forget to do it with my own modules.  Something like [this](https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/exploit/windows/local/ms16_016_webdav.md)

Also, delete all of the remnant binaries in `external/source/exploits/CVE-2016-0040`.  It looks like some may have been leftover from renaming that aren't tracked by the solution file.  Other than that, everything looks good so far.

**Comment by de7ec7ed** [comment]
Beyond module docs and removing files is there anything else that needs to be changed? It took me a while to get back to the last set of changes and I don't want this to hang out open for to long on you guys from my 1 week+ turn around times.

**Comment by wwebb-r7** [comment]
If I knew the answer to that, I'd have told you the first time.

All I can say is don't worry about putting us out with regards to turnaround times.  This thing is still young as far as pull requests go.

**Comment by de7ec7ed** [comment]
Sounds good, I guess I was hoping you would point me to some checklist. Not a problem though and thanks for the help.

**Comment by busterb** [comment]
I think that's it. We only recently added the documentation step to the pull request template checklist.

**Comment by bwatters-r7** [comment]
FYI....
image

I think this matches the expected outcome.

**Comment by wwebb-r7** [comment]
@de7ec7ed just pinging you on module docs and binary cleanup, looks like we're ready to merge otherwise

**Comment by wwebb-r7** [comment]
Retested recently due to inactivity.  At this point I'll just correct it myself.

**Comment by de7ec7ed** [comment]
Sorry for the long delay. I will try to get to the docs by the end of next week.

**Comment by de7ec7ed** [comment]
@wwebb-r7 I think I have addressed the feedback now.

File: documentation/modules/exploit/windows/local/ms16_014_wmi_recv_notif.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+## Intro
+
+This module exploits a uninitialized stack variable vulnerability present in the WMI subsystem  of `ntoskrnl`.  The vulnerability is described by MS16-014 and CVE-2016-0040.  The module allows the user to elevate privileges (NT AUTHORITY\SYSTEM)
```
Comment (h00die) [inline documentation/modules/exploit/windows/local/ms16_014_wmi_recv_notif.md]: without looking at the code, does this only work against win7 (due to offsets etc) or is it only TESTED against win7.  The MS bulletin showed vista and 2008 and many others vulnerable.

Reason I ask, is that this intro is a good place to add something like:
This module has been tested against the following systems:
- Windows 7 x64 sp1
- Foobarbaz

This module should work, but has not been tested against:
- Windows 2008


That way a user will know what it definitely works for, and may (to make their own risk assessment).

File: documentation/modules/exploit/windows/local/ms16_014_wmi_recv_notif.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+## Usage
+
+You'll first need to obtain a session on the target system.  Next, once the module is loaded, one simply needs to set the `payload` and `session` options.  From here, running the module will result in the payload being executed with system level privileges.
+
+An example session follows:
```
Comment (h00die) [inline documentation/modules/exploit/windows/local/ms16_014_wmi_recv_notif.md]: Change this to be a 3rd level heading with the OS name, version and architecture (`### Windows 7 SP? x64`)

**Review by h00die — COMMENTED** [review]
doc format updates

**Comment by de7ec7ed** [comment]
I'm sorry I used the previously referenced link (https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/exploit/windows/local/ms16_016_webdav.md) to create the documentation page. I'll update it with your new requirements soon.

**Comment by bwatters-r7** [comment]
@de7ec7ed any timeline on this?

**Comment by bwatters-r7** [comment]
@de7ec7ed pinging you again......

**Comment by mubix** [comment]
@bwatters-r7 what work is left? If it's just a few documentation changes I would think that the module being in master plays more than waiting for a first time contributor to complete a bunch of documentation. If it's something I can help complete please let me know. Looking through the modudle and documentation now.

**Comment by bwatters-r7** [comment]
@mubix I have done very little with it.  The original person assigned to it is no longer a committer, and I grabbed it as I have a test infrastructure for it.  I notice that @h00die did a bunch of doc suggestions which is awesome, but left is to (1) fix some syntactical issues in the ruby code you can see in rubocop, (2) verify the C code, (3) recompile from source and sign the new binary, and (4) retest.  I had placed this on the back-burner because it looked like @de7ec7ed had something coming.  I'd give him first priority to push stuff out, but otherwise, I was going to hit it after finishing up some internal stuff and the peinject PR.

**Comment by bwatters-r7** [comment]
I went ahead and cleaned this up a bit and landed it.

**Comment by bwatters-r7** [comment]
# Release notes
This PR adds a local priv esc targeting Windows 7x64 sp0 and sp1 with CVE-2016-0040.

**Pull request merged**